prboom2/src/SDL/i_system.c

```diff
@@ -285,24 +285,41 @@ const char* I_GetTempDir(void)
 }
 
 #else
-// cph - V.Aguilar (5/30/99) suggested return ~/.lxdoom/, creating
-//  if non-existant
-// cph 2006/07/23 - give prboom+ its own dir
-static const char prboom_dir[] = {"/.dsda-doom"}; // Mead rem extra slash 8/21/03
+
+static char *I_ConcatDir(const char *var, const char *subdir)
+{
+  static char buffer[PATH_MAX];
+  size_t len;
+
+  // Remove trailing slashes to be safe.
+  strncpy(buffer, var, sizeof(buffer));
+  len = strlen(buffer);
+  while (len > 0 && buffer[len - 1] == '/')
+    buffer[--len] = 0;
+  // Append subdirectory.
+  strncpy(&buffer[len], subdir, sizeof(buffer) - len);
+
+  return buffer;
+}
 
 const char *I_DoomExeDir(void)
 {
   static char *base;
   if (!base)        // cache multiple requests
   {
     char *home = M_getenv("HOME");
-    size_t len = strlen(home);
-
-    base = Z_Malloc(len + strlen(prboom_dir) + 1);
-    strcpy(base, home);
-    // I've had trouble with trailing slashes before...
-    if (base[len-1] == '/') base[len-1] = 0;
-    strcat(base, prboom_dir);
+    // First, try legacy directory.
+    base = I_ConcatDir(home, "/.dsda-doom");
+    if (access(base, F_OK) != 0)
+    {
```

Comment:
Once I_ConcatDir allocates its response, we'll need to Z_Free it here before getting the new value.